src/pkg/podman/podman.go

```diff
@@ -226,18 +237,80 @@ func IsToolboxImage(image string) (bool, error) {
 	return true, nil
 }
 
-// Pull pulls an image
-func Pull(imageName string) error {
+func Login(registry string, extraArgs ...string) error {
+	var stderr bytes.Buffer
+
 	logLevelString := LogLevel.String()
-	args := []string{"--log-level", logLevelString, "pull", imageName}
+	args := []string{"--log-level", logLevelString, "login", registry}
+	args = append(args, extraArgs...)
 
-	if err := shell.Run("podman", nil, nil, nil, args...); err != nil {
+	if err := shell.Run("podman", os.Stdin, os.Stdout, &stderr, args...); err != nil {
+		if stderr.Len() == 0 {
+			return fmt.Errorf("failed to login to %s: %w", registry, err)
+		}
+
+		err := parseErrorMsg(&stderr).(*internalError)
 		return err
 	}
 
 	return nil
 }
 
+// Pull pulls an image. Wraps around command 'podman pull'.
+//
+// The image pull can fail for many reasons. The recognized reasons are:
+//   - manifest unknown (ErrManifestUnknown)
+//   - name unknown (ErrNameUnknown)
+//   - unauthorized (ErrUnauthorized)
+//
+// To learn more about OCI API error codes, see: https://github.com/opencontainers/distribution-spec/blob/main/spec.md
+//
+// The wrapper tries to discern these errors to some extent and provide means
+// for their handling. The provided errors can not be relied on completely as
+// they are created based on error message parsing that differs based on used
+// registry.
+func Pull(imageName string) error {
+	var stderr bytes.Buffer
+
+	logrus.Debugf("Pulling image %s", imageName)
+
+	logLevelString := LogLevel.String()
+	args := []string{"--log-level", logLevelString, "pull", imageName, "--quiet"}
+
+	if err := shell.Run("podman", nil, nil, &stderr, args...); err != nil {
+		if stderr.Len() == 0 {
+			return fmt.Errorf("failed to pull image %s: %w", imageName, err)
+		}
+
+		err := parseErrorMsg(&stderr).(*internalError)
+
+		// This error is returned when the manifest, identified by name and tag
+		// is unknown to the repository.
+		if err.Is(errors.New("manifest unknown")) {
+			return ErrManifestUnknown
+		}
+
+		// This is returned if the name used during an operation is unknown to the
+		// registry.
+		// The "unknown" error message is an exception applying to registry.redhat.io
+		if err.Is(errors.New("name unknown")) || err.Is(errors.New("unknown")) {
+			return ErrNameUnknown
+		}
+
+		// The access controller was unable to authenticate the client. Often this
+		// will be accompanied by a Www-Authenticate HTTP response header
+		// indicating how to authenticate.
```

Comment:
This comment seems redundant, unless we can somehow leverage the HTTP response header to our benefit. Otherwise, the error message makes it clear that this is about authentication.

Reply:
Removed the part mentioning the response header.